Fix(#12318): Fixes a bug in evaluation of JSONPath for wildcard and s…

marcodallasanta opened the pull request:
## Motivation
This is an attempt to fix issue #12318 

## Changes
Changes jsonpath.py. 
When no matches, then it checks if the pattern of the JSONPath contains an array slice or wildcard, if so it gracefully returns an empty array instead of failing with NoSuchJsonPathError.

There's an exception though, if the JSONPath starts with an array slice syntax than it should fail.

## Testing
There's a parameterised test for each of the pattern I identified to be problematic, but below there's a little script if it can help in diagnose issues

```/bin/bash
# 1. start docker
docker run --rm -d --name "localstack-main" -p 4566:4566 localstack/localstack

# 2. create a state machine - Change the JsonPath as you like
stateMachineArn=$(awslocal stepfunctions create-state-machine --name "test" --definition '{
  "StartAt": "Pass",
  "States": {
    "Pass": {
      "Type": "Pass",
      "End": true,
      "Parameters": {
        "types.$": "$.items[*]"
      }
    }
  }
}'  --role-arn "dummy" | jq -r ".stateMachineArn");

# 3. create an execution 
executionArn=$(awslocal stepfunctions start-execution --state-machine-arn "$stateMachineArn" | jq -r ".executionArn");

# 4. describe the execution (jq for syntax highlight)
awslocal stepfunctions describe-execution --execution-arn "$executionArn" | jq "."
```

marcodallasanta commented:
I wanted to add a label as `semver: patch`, but looks like I don't have permissions.

marcodallasanta commented:
I have read the CLA Document and I hereby sign the CLA

marcodallasanta commented:
I have read the CLA Document and I hereby sign the CLA

marcodallasanta commented:
recheck

MEPalma commented on tests/aws/services/stepfunctions/v2/base/test_base.snapshot.json:
```diff
@@ -1396,5 +1374,821 @@
       }
     }
+  },
+  "tests/aws/services/stepfunctions/v2/base/test_base.py::TestSnfBase::test_json_path_array_wildcard_or_slice_with_no_input[$[*]]": {
```
Can we revert the changes in this file to only include the additions of your new test case `test_json_path_array_wildcard_or_slice_with_no_input`

Please, also track the changes for `test_json_path_array_wildcard_or_slice_with_no_input` in the `test_base.snapshot.json` file.

MEPalma reviewed (commented):
Thank you @marcodallasanta for contributing to LocalStack and providing a solution for https://github.com/localstack/localstack/issues/12318! Your proposed approach will definitely help unblock this scenario.

I just have a minor refinement request regarding the snapshot tests. Specifically, could you please update your PR to ensure that only the snapshot information relevant to your new test case is included? Additionally, please make sure to incorporate your test case changes into the validation file.

For reproducibility, kindly avoid manually editing the generated files. Instead, use transformers within the `sfn_snapshot` instance if necessary (though this should not be required in this case).

If you’d prefer, I’d also be happy to take care of these changes—just let me know!

Thanks again, looking forward to merging this soon!

marcodallasanta commented:
Thank you @MEPalma, will try myself just to gain a more experience with the repo in case I will contribute again, but if it takes too long I will leave it to you.

marcodallasanta commented:
Hey @MEPalma,  I did what you suggested above: 
- reverted and updated only with the relevant tests the snapshot file.
- updated the validation.json file.

Let me know if there's anything else that needs to be done before merging.
